_posts/2021-01-30-Writing-Better-Code-Cloud-Code-with-Parse-Server-Example.md

```diff
@@ -0,0 +1,191 @@
+
+Using Promise.all:
+```javascript
+Parse.Cloud.define('testFunction', ({user, params:{name, type}}) => {
+  const nameQuery = new Parse.Query('Name');
+  nameQuery.equalTo('name', name);
+  const typeQuery = new Parse.Query('Type');
+  typeQuery.equalTo('type',type);
+  const [nameData, typeData] = await Promise.all([nameQuery.first({useMasterKey:true}),typeQuery.first({useMasterKey:true})]);
+  if (!nameData) {
+    throw new Parse.Error(141, 'name does not exist');
+  }
+  if (!typeData) {
+    throw new Parse.Error(141, 'type does not exist');
+  }
+  user.set('type', type);
+  user.set('name', name);
+  await user.save(null, {useMasterKey: true});
+  return 'Type and name are valid'
+},{
+  requireUser:true,
+  fields: ['name','type']
+});
+```
+
+These are only a few examples of how to write better cloud code. On the latest [Parse Server Example](https://github.com/parse-community/parse-server-example) project, you can now run `npm run prettier` to style your code, followed by `npm run lint` , which will run over your cloud code and check for quality issues. You can also run `npm run lint-fix` to fix any easy errors, such as indentation.
+
+# Testing
+
+As your Parse Server project gets larger and more complex, it can be challenging to add new features without previous ones becoming affected.
+
+Automated testing is a good way to be assured that your functions are running as expected.
+
+On the latest [Parse Server Example](https://github.com/parse-community/parse-server-example) project, you can create tests by creating `*.spec.js` files in `/spec`. 
+
+This will boot up a mock database, start a Parse Server based off your `index.js`. You don't have to worry about tests impacting your production database.
+
+We have written some example tests to show you how it works.
+
+Now, run `npm run test` and `jasmine` (the testing framework) will check to make sure the tests run as expected. 
+
+Some helpful tips:
+- If you create a test using `fit()` or `fdescribe()`, jasmine will focus on those tests only, and others will be ignored.
+- You can specify a file to test by using `npm run test /spec/testFile.spec.js`
+- You won't be able to access any live data through the tests. You'll have to save mock data using `beforeAll()` if you want to test queries.
+
+# Coverage
+
+Now you've learnt testing, it's important to make sure as much of your cloud functions are covered by tests, so you can be sure that they are all running as expected.
+
+After you've written and tested a few tests, you can run `npm run coverage`. This will create a new folder `/coverage`. Navigate to `/coverage/lcov-report` and open `index.html`, and you'll be able to visually see what bits of your code you need to write tests for, and which sections are covered. The more of your cloud code you cover, the more you can be assured that changes don't have any unforseen consequences.
```

Comment:
```suggestion
After you've written and tested a few tests, you can run `npm run coverage`. This will create a new folder `/coverage`. Navigate to `/coverage/lcov-report` and open `index.html`, and you'll be able to visually see what bits of your code you need to write tests for, and which sections are covered. The more of your Cloud Code you cover, the more you can be assured that changes don't have any unforeseen consequences.
```